### ISSUE TITLE
Support for recording clicks element wise

### ISSUE THREAD

**Issue by pivotal-casebook** [facebookarchive/huxley#29]
Hi!

We were tinkering with Huxley and wanted to know if huxley supports/would support recording clicks on a webpage during a test on an element basis instead of cursor position.

This might be helpful to avoid rerecording the test when the position of a UI element on a webpage changes. We would need to rerecord the test  because the old click position might not be over the UI element in the new layout. If the recording was done on element basis, the clicking would always work irrespective of the position of the element. Right now, huxley --playback-only fails because it cannot find a ui element at that position with the following error,

"raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.WebDriverException: Message: u"document.elementFromPoint(0, 700) is null\nBuild info: version: '2.35.0', revision: 'c916b9d', time: '2013-08-12 15:42:01'\nSystem info: os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.8.2', java.version: '1.6.0_37'\nDriver info: driver.version: unknown" ; Screenshot: available via screen ;"

This would be helpful to avoid large amount test rerecording in a large test suite when there are small UI changes.

Please let us know your views.

Thanks,

**Comment by petehunt** [comment]
The original (pre-open source) version of Huxley worked like this. In fact it's how Facebook writes webdriver tests (i.e. "click an element with this xpath"). The problem is that when you change the implementation details of your UI due to refactoring markup/JS/etc you may break the test and force a re-record of everything. I've found that using x/y has been pretty resilient to both UI and markup changes. Have you identified a particular case that this fails at (i.e. you say small UI changes, can you give an example?) I would be willing to make this strategy pluggable, though it would take a bit of refactoring.

**Comment by pivotal-casebook** [comment]
An example would be, suppose I want to click on an menu item in a menu bar. If the UI changes the position of that particular menu item within that menu bar, my click event(on the basis of position) will no longer click on the menu item. If this menu item is used at numerous places, I would need to rerecord the tests(including interactions) for all those pages.

If we do click event by element, all I would need to do to update the test suite is to update the screenshots by running the huxley command again. I would not have to worry about recording the interactions again( because they would remain the same in the record.json file since they are stored element wise)

**Comment by petehunt** [comment]
So the x/y clicking is supposed to simulate a human running through the test plan (in an automated way). I think what you want is actually clicking on the menu item with the same text each time, right?

What if instead Huxley looked at the item that was clicked and looks for a data-huxley-id attribute, which will be used instead of x/y if it exists?

**Comment by pivotal-casebook** [comment]
Yeah. exactly. It would be nice if Huxley looked at the item that was clicked through a variable like data-huxley-id attribute.

However, I'm wondering why we would need to maintain another attribute (data-huxley-id), instead of using the existing element id?  

Thanks

**Comment by petehunt** [comment]
It's extremely important that Huxley is resilient to changes in the underlying implementation, so I would like to expose this as an explicit contract between Huxley and the application being tested rather than have it be implicit, magical, and hard to understand.

**Comment by pivotal-casebook** [comment]
got it! That would be an excellent feature. Thanks!

**Comment by pivotal-casebook** [comment]
Hey!
We were wondering if you have plans to implement this in the near future.
